### PR TITLE
simpler diametrizer

### PR DESCRIPTION
**Pull request by arnaudon**

### CODE REVIEW

**Comment by arnaudon** [comment]
@adrien-berchet , also when you have time!

**Comment by arnaudon** [comment]
sorry, it's a huge mess, because the test had missing `assert compar_dicts`, so the input data was not tested and was wrong on disk, then the diamtrized test were also wrong wrt correct input data. I also changed a few things like basal-> basal_dendrite to avoid using the dict mapping

**Comment by arnaudon** [comment]
@adrien-berchet , did you get a chance to have a look at that? No big rush, just to know/plan work.

**Comment by adrien-berchet** [comment]
> @adrien-berchet , did you get a chance to have a look at that? No big rush, just to know/plan work.

Ah no, I forgot about this one, sorry. I will very soon!

File: examples/get_morphologies.py [hunk 1]
```diff
@@ -18,13 +18,25 @@
...
 
 
+def get_url(url):
+    """Open a URL without SSL verification."""
+    ctx = ssl.create_default_context()
+    ctx.check_hostname = False
+    ctx.verify_mode = ssl.CERT_NONE
+    ctx.set_ciphers("DEFAULT@SECLEVEL=1")
+    req = Request(url)
+    response = urlopen(req, context=ctx)  # pylint: disable=consider-using-with
```
Comment (adrien-berchet) [inline examples/get_morphologies.py:36]: Why not using a context manager as suggested by pylint?

File: examples/run.sh [hunk 2]
```diff
@@ -1,10 +1,10 @@
 #!/bin/bash
 
 # fetch some morphologies from neuromorpho.org
-python get_morphologies.py
+#python get_morphologies.py
```
Comment (adrien-berchet) [inline examples/run.sh:4]: Maybe you can run it only id the `morphology` folder does not exist?

File: tests/data/simpler_model_data.json [hunk 3]
```diff
@@ -0,0 +1 @@
...8017690335927, 0.10092123989396076, 0.04680877854080963, 0.03134206997162865, 0.7827281142242415, 1.0, 0.7721448184675089, 0.9855848859679315, 0.9534542307119734, 0.6962178219191657, 0.7640790516447591, 0.8660723409534996, 0.032019353948400074, 0.855147949386722], "apical_dendrite": [1.0, 0.7443870208169504, 0.3647110695224464, 0.12184497718515643, 0.3516306644940961, 0.7126552920188522, 0.017892523333600956, 0.006267798332285662, 0.8881232630075685]}, {"basal_dendrite": [3.0850000381469727, 0.9300000071525574, 0.6642856597900391, 0.4165624976158142, 0.31000006198883057, 0.9300000071525574, 0.7045454382896423, 0.7233332991600037, 0.7233333587646484, 1.1119999885559082, 0.7733333110809326, 0.9300000071525574, 0.4871428310871124, 1.2333332300186157, 0.3752632141113281, 0.8524999618530273, 0.6842105388641357, 0.6197144389152527, 0.15000000596046448, 0.9300000071525574, 0.9300000071525574, 0.33268293738365173], "apical_dendrite": [2.680000066757202, 1.7899999618530273, 2.0899999141693115, 1.6699998378753662, 0.32999998331069946, 0.5162963271141052, 0.30000001192092896, 0.6000000238418579, 0.3656250238418579]}, {"basal_dendrite": 6.981783576001451, "apical_dendrite": 5.776746786628541}]
```
Comment (adrien-berchet) [inline tests/data/simpler_model_data.json:1]: Don't forget to add new lines at the end of each file

File: tests/data/model_params.json [hunk 4]
```diff
@@ -1,160 +1,128 @@
...
-                        "a": 4,
-                        "loc": -0.0031,
-                        "max": 0.0637,
-                        "min": 0.0,
-                        "num_value": 294,
-                        "scale": 0.0285
-                    },
-                    "sequential": null
-                }
+    "sibling_ratios": {
```
Comment (adrien-berchet) [inline tests/data/model_params.json:2]: Why did you change these values? Changing them also changes expected values in tests, so it's harder to know if the new expected values changed because of changes in inputs or in the code.

File: diameter_synthesis/build_diameters.py [hunk 5]
```diff
@@ -30,23 +30,12 @@
...
-    "apical": SectionType.apical_dendrite,
-    "axon": SectionType.axon,
-}
-
-TYPES_TO_STR = {
-    SectionType.basal_dendrite: "basal",
-    SectionType.apical_dendrite: "apical",
-    SectionType.axon: "axon",
-}
-
```
[the remark this answers is not in the document]
Comment (arnaudon) [reply diameter_synthesis/build_diameters.py:49]: yes, maybe it's to much, let me check how this is used view neurots

File: diameter_synthesis/build_models.py [hunk 6]
```diff
@@ -129,32 +127,43 @@ def extract_parameters(sampling_model, morphologies, config):
         all_data["trunk_diameters"][neurite_type] = []
         all_data["tapers"][neurite_type] = []
 
+        def _float(data):
+            return [float(d) for d in data]
+
```
[the remark this answers is not in the document]
Comment (arnaudon) [reply diameter_synthesis/build_models.py:132]: I think this is for json if I recall

File: diameter_synthesis/main.py [hunk 7]
```diff
@@ -67,23 +69,31 @@ def plot_models(morphologies, config, models_params, models_data, ext=".png"):
...
-                fig_name = Path(config["fig_folder"]) / mtype / fit_tpe
-                plot_distribution_fit(
-                    model_data[fit_tpe],
-                    model_param[fit_tpe],
-                    config["neurite_types"],
-                    fig_name=fig_name,
-                    ext=ext,
-                )
+    for model in config["models"]:
+        if len(models_data[model][list(morphologies.keys())[0]]) == 3:
```
[the remark this answers is not in the document]
Comment (arnaudon) [reply diameter_synthesis/main.py:73]: maybe, but this is just plotting stuff

File: diameter_synthesis/main.py [hunk 8]
```diff
@@ -67,23 +69,31 @@ def plot_models(morphologies, config, models_params, models_data, ext=".png"):
...
+                    os.mkdir(Path(config["fig_folder"]) / mtype)
+                for fit_tpe in fit_tpes:
+                    fig_name = Path(config["fig_folder"]) / mtype / fit_tpe
+                    plot_distribution_fit(
+                        model_data[fit_tpe],
+                        model_param[fit_tpe],
+                        config["neurite_types"],
+                        fig_name=fig_name,
+                        ext=ext,
+                    )
```
[the remark this answers is not in the document]
Comment (arnaudon) [reply diameter_synthesis/main.py:96]: nah, this is plotting stuff

File: diameter_synthesis/simpler_diametrizer.py [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+
+
+def _map_sections(fun, neurite, iterator_type=Section.ipreorder):
+    """Map `fun` to all the sections."""
+    return list(map(fun, iterator_type(neurite.root_node)))
+
+
+def terminal_path_lengths(neurite, cache):
+    """Get the path lengths to each terminal point."""
+    return _map_sections(partial(section_path_length, cache=cache), neurite, Section.ileaf)
```
[the remark this answers is not in the document]
Comment (arnaudon) [reply diameter_synthesis/simpler_diametrizer.py:34]: yes, it improves massively, due to recursive algo I use to trasverse.

File: examples/get_morphologies.py (hunk 1 above)
Comment (arnaudon) [reply examples/get_morphologies.py:36]: I don't know, I can add that yes

File: examples/run.sh (hunk 2 above)
Comment (arnaudon) [reply examples/run.sh:4]: not sure why I commented this lol

File: tests/data/model_params.json (hunk 4 above)
Comment (arnaudon) [reply tests/data/model_params.json:2]: these are the expected values to test against, no? There was a bug in t he test, you didn't add `assert` to compare things, so half of the test where outdated, so these large changes

File: tests/data/simpler_model_data.json (hunk 3 above)
Comment (arnaudon) [reply tests/data/simpler_model_data.json:1]: ah why?

File: tests/test_build_models.py [hunk 10]
```diff
@@ -23,33 +23,36 @@
...
+        "neurite_types": ["basal_dendrite", "apical_dendrite"],
         "terminal_threshold": 2,
         "taper": {"max": 1e-6, "min": -0.01},
     }
     res_astrocyte = build_models.build(single_pop, config_astrocyte, with_data=True)
 
     assert len(res_astrocyte) == 2
-    compare_dicts(res_astrocyte, [model_params, model_data], precision=3)
+
+    assert compare_dicts(res_astrocyte, [astro_model_params, astro_model_data], precision=3)
```
Comment (arnaudon) [inline tests/test_build_models.py:55]: @adrien-berchet , these asserts

File: diameter_synthesis/build_diameters.py [hunk 11]
```diff
@@ -361,6 +350,7 @@ def _diametrize_neuron(params_tree, neuron, params, neurite_types, config, rng=n
                     L.warning("max tries attained with %s", neurite_type)
                     wrong_tips = False
                 n_tries += 1
+    return neuron
```
[the remark this answers is not in the document]
Comment (arnaudon) [reply diameter_synthesis/build_diameters.py:353]: it's like in the previous version, no? but yes, I can just leave it in place and not return it

File: examples/get_morphologies.py (hunk 1 above)
Comment (arnaudon) [reply examples/get_morphologies.py:36]: it's the same as in morphology-workflow example, I don't care, I'll leave it as is

File: diameter_synthesis/build_diameters.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (arnaudon) [reply diameter_synthesis/build_diameters.py:353]: i remove the returns

File: tests/data/simpler_model_data.json (hunk 3 above)
Comment (arnaudon) [reply tests/data/simpler_model_data.json:1]: done

File: diameter_synthesis/build_models.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (adrien-berchet) [reply diameter_synthesis/build_models.py:132]: Ah yes probably.

File: diameter_synthesis/main.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (adrien-berchet) [reply diameter_synthesis/main.py:73]: image

:laughing:

File: diameter_synthesis/main.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (adrien-berchet) [reply diameter_synthesis/main.py:96]: lol

File: diameter_synthesis/simpler_diametrizer.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (adrien-berchet) [reply diameter_synthesis/simpler_diametrizer.py:34]: Ok, I see. So we can keep it there for now but it would be very nice to submit such feature to NeuroM.

File: examples/get_morphologies.py (hunk 1 above)
Comment (adrien-berchet) [reply examples/get_morphologies.py:36]: Yeah ok, I don't really know the point of a context manager in this context.

File: examples/run.sh (hunk 2 above)
Comment (adrien-berchet) [reply examples/run.sh]: :smile:

File: tests/test_build_models.py (hunk 10 above)
Comment (adrien-berchet) [reply tests/test_build_models.py:55]: :scream: :facepalm:

File: tests/data/model_params.json (hunk 4 above)
Comment (adrien-berchet) [reply tests/data/model_params.json:2]: Oh boy :facepalm: 
Ok, I understand why you changed :-)

File: diameter_synthesis/simpler_diametrizer.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (arnaudon) [reply diameter_synthesis/simpler_diametrizer.py:34]: go for it!

File: setup.py [hunk 12]
```diff
@@ -39,6 +39,7 @@
     "sphinx-bluebrain-theme",
     "sphinx-jsonschema",
     "sphinx_click",
+    "docutils<0.19",  # related to https://github.com/CrossNox/m2r2/issues/52
```
Comment (adrien-berchet) [inline setup.py:42]: The new version that fixes this was released 45min after you merged this :joy:
